### PR TITLE
feat: adding new `abi-typegen` package

### LINKED ISSUE #360: fix: Enums, structs nested in tuples and arrays should infer names from type
`__tuple_element` and `__array_element` are used in the components.   The type fields have the preserve the names; this field should be used to create namespaces.

- add more tests to prevent regressions e.g. nested enums and structs
- handle potential naming collisions

### LINKED ISSUE #478: Typechain generated files with problems in typing
### What version of fuels-ts are you using?

0.12.0

### Steps to Reproduce

1- in `master` branch and root folder, run `pnpm --filter typechain-target-fuels build:example`
2- go to `packages/typechain-target-fuels/example/types/MulticallAbi.d.ts`, the error is there:
```
Interface 'MulticallAbiInterface' incorrectly extends interface 'Interface'.
  Types of property 'encodeFunctionData' are incompatible.
    Type '(functionFragment: "main", values: [ScriptDataInput]) => Uint8Array' is not assignable to type '(functionFragment: string | FunctionFragment, values: InputValue[]) => Uint8Array'.
      Types of parameters 'values' and 'values' are incompatible.
        Type 'InputValue[]' is not assignable to type '[ScriptDataInput]'.
          Target requires 1 element(s) but source may have fewer.ts(2430)
```

### Expected Behavior

Typing should be correct

### Actual Behavior

I guess the problem is with new `undefined` type that came with `Option` support

### LINKED ISSUE #480: Simplify type generation CLI
Today users need to install `fuelchain` and `typechain-target-fuels` to generate Types from a `ABI JSON`. The idea is to implement a single CLI tool for type generation and remove the need to install `typechain-target-fuels`.

### LINKED ISSUE #520: Refactor Interface, Contact, etc to use Flat ABI
### Motivation

Currently, we allow both nested and flat ABI.  However, this is causing our code to be harder to maintain.  Going forward, we should support the Flat ABI and refactor the code path.

The following are types that can be simplified.

- [ ] `ParamType`
- [ ] `JsonFragment`
- [ ] `FunctionFragment`

### Usage example

_No response_

### Possible implementations

_No response_

### LINKED ISSUE #544: Fix type generation wrong for Vec inputs
After generating types from a contract that uses `Vec<b256>`, the code generated creates a Vec Structure instead of the correct type like in the case of `Vec<b256>` as input of type `Array<string>`.

Examples:
```abi
fn echo_b256_vector(vector_1: Vec<b256>, index: u32) -> b256;
```

Types generated
```ts
  encodeFunctionData(
    functionFragment: "echo_b256_vector",
    values: [VecInput, BigNumberish]
  ): Uint8Array;
```

Expected Types generated
```
  encodeFunctionData(
    functionFragment: "echo_b256_vector",
    values: [Array<string>, BigNumberish]
  ): Uint8Array;
```

### LINKED ISSUE #569: Fix type generation wrong for Array inputs
All types inside arrays are currently lost.

More info here:
 - https://github.com/FuelLabs/fuels-ts/issues/544#issuecomment-1293577816_

### PR DESCRIPTION
**Pull request by arboleya**
# Summary
 - [x] Close #360
 - [x] Close #478
 - [x] Close #480 
 - [x] Close #520
 - [x] Close #544 
 - [x] Close  #569

> **Note**: _This PR needs to be merged after_ #612.
>  1. it will inherit the new `abi-coder`/`b512` type from there
>  1. it should review and update docs for generating types

# TL;DR

Before:
```console
npm install fuels --save
npm install fuelchain typechain-target-fuels --save-dev
npx fuelchain --target=fuels --out-dir=./src/contracts ../contract/out/debug/*-abi.json
```

Now:
```console
npm install fuels --save
npx fuels typegen -i ../contract/out/debug/*-abi.json -o ./src/contracts
```

# Description

This PR adds a new `abi-typegen` package.

Key features:
 - Heavily tested
 - Dependency-free
 - Tailored for Sway ABI

It is a drop-in replacement for the old _[removed]_ packages:
 - `fuelchain`
 - `typechain-target-fuels`

# Testing out

 1. Switch to this branch
 2. `pnpm install && pnpm build`
 3. `pnpm -r exec pnpm link --global --dir ./` _([docs](https://github.com/FuelLabs/fuels-ts/blob/master/docs/CONTRIBUTING.md#on-fuels-ts-root-directory))_
 4. `npx fuels typegen -i ../contract/out/debug/*-abi.json -o ./src/contracts`

# Docs

  - [x] `fuel-ts`
    - [README.md](https://github.com/FuelLabs/fuels-ts/blob/aa/typegen-revamp/README.md#generate-contract-types-from-abi)
    - `docs/`[QUICKSTART.md](https://github.com/FuelLabs/fuels-ts/blob/aa/typegen-revamp/docs/QUICKSTART.md#generating-contract-types)
  - [x] `fuel-docs`:
    - `https://github.com/FuelLabs/fuel-docs/pull/47

# Notes

 - [ ] Remember to merge the documentation's [PR](https://github.com/FuelLabs/fuel-docs/pull/47) on `fuel-docs`
 - [x] Remember to review/update docs after rebasing on top of:
   - #612

### CODE REVIEW

File: packages/abi-typegen/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# `@fuel-ts/abi-coder`
```
Comment (luizstacio) [inline packages/abi-typegen/README.md:1]: Package name should be abi-typegen

File: packages/abi-typegen/package.json [hunk 2]
```diff
@@ -0,0 +1,50 @@
+{
+  "name": "@fuel-ts/address",
```
Comment (luizstacio) [inline packages/abi-typegen/package.json:2]: Package name is wrong here also it should be: `abi-typegen`

File: packages/abi-typegen/src/abi/abi.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    /*
+      First we format all attributes
+    */
+    const fnsTypedefs = this.functions.map((f) => f.getDeclaration());
+
+    const fnsFragments = this.functions.map((f) => `${f.attributes.name}: FunctionFragment;`);
+
+    const encoders = functions.map((f) =>
+      DTS_TEMPLATE_ENCODER.replace('{NAME}', f.attributes.name).replace(
```
Comment (luizstacio) [inline packages/abi-typegen/src/abi/abi.ts:83]: I liked the idea of the DTS_TEMPLATE_ENCODER maybe, we should also create a wrapper function `templateEncoder(name: string, input: Input): string` this function could be inside the templates.

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/abi-typegen/src/templates/factory.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+export const FACTORY_TEMPLATE = `export /* Autogenerated file. Do not edit manually. */
+/* tslint:disable */
+/* eslint-disable */
+
+import type { Provider, Wallet, AbstractAddress } from "fuels"
```
Comment (luizstacio) [inline packages/abi-typegen/src/templates/factory.ts:5]: The Wallet was update to BaseWalletLocked

File: packages/abi-typegen/package.json [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    "yargs": "^17.6.0"
+  },
+  "devDependencies": {
+    "@types/glob": "^8.0.0",
+    "@types/lodash.upperfirst": "^4.3.7",
+    "@types/mkdirp": "^1.0.2",
+    "@types/node": "^18.11.7",
+    "@types/rimraf": "^3.0.2",
+    "@types/yargs": "^17.0.13",
+    "ts-node": "^10.9.1",
```
Comment (luizstacio) [inline packages/abi-typegen/package.json:43]: I think ts-node is not been used

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/abi-typegen/src/abi/abi.ts (hunk 3 above)
Comment (arboleya) [reply packages/abi-typegen/src/abi/abi.ts:83]: Are you talking about these wrappers?
 - `encodeFunctionData`
 - `decodeFunctionData`

https://github.com/FuelLabs/fuels-ts/blob/e6e6890e9f11dc9323917101aa4e7f367ab0bb5d/packages/abi-typegen/src/templates/dts.ts#L37-L39

Comment (luizstacio) [reply packages/abi-typegen/src/abi/abi.ts:83]: Like this;
```ts
export const DTS_TEMPLATE_DECODER = `decodeFunctionData(functionFragment: '{NAME}', data: BytesLike): DecodedValue`;

export function templateDecoder(name: string, inputs: IAbiFunctionIO[]) {
    return DTS_TEMPLATE_ENCODER.replace('{NAME}', f.attributes.name).replace('{INPUT}', inputs)
```

File: packages/abi-typegen/test/fixtures/array-nested.sw [hunk 6]
```diff
@@ -0,0 +1,31 @@
+contract;
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/test/fixtures/array-nested.sw:1]: Do you have a sway example to illustrate your rationale?

File: packages/abi-typegen/src/utils/getPackageVersion.test.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import packageJson from '../../package.json';
+
+import { getPackageVersion } from './getPackageVersion';
+
+describe('getPackageVersion.ts', () => {
+  test('should get package version just fine', () => {
+    const expectedVersion = packageJson.version;
+    expect(getPackageVersion().version).toEqual(expectedVersion);
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/src/utils/getPackageVersion.test.ts:8]: Here I am looking for the `package.json` inside the `abi-typegen` package, not the root one.

If we use the `fuels` one, it will fail when installing the package in a standalone fashion. Right?

File: packages/abi-typegen/src/templates/index.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+  const dtsBuffer: string[] = [];
+  const factoriesBuffer: string[] = [];
+
+  params.abis.forEach((abi) => {
+    dtsBuffer.push(`export type { ${abi.name} } from './${abi.name}';`);
+    factoriesBuffer.push(`export { ${abi.name}__factory } from './factories/${abi.name}__factory'`);
+  });
+
+  const output = `/* Autogenerated file. Do not edit manually. */
+/* Fuels version: ${params.abis[0].fuelsVersion} */
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/src/templates/index.ts:13]: Good idea, but requires attention — it should work when installing the package via `fuels` or directly via `@fuel-ts/abi-typegen`.

File: packages/abi-typegen/src/types/U8Type.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  public static swayTypeExample = 'u8';
+
+  public name = 'u8';
+
+  public static MATCH_REGEX: RegExp = /^u8$/m;
+
+  constructor(params: { rawAbiType: IRawAbiTypeRoot }) {
+    super(params);
+    this.attributes = {
+      inputLabel: `BigNumberish`,
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/src/types/U8Type.ts:16]: Yes. I got this `BigNumberish` x `number` from the previous version, as per @luizstacio instructions.

I'm afraid I didn't get your suggestion, though.

What do you think I should change here?

File: packages/abi-typegen/README.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+| ------------------: | :----------------------------- | :---------------------------------: | :---------------------------------: |
+|              **u8** | `255`                          |           `BigNumberish`            |              `number`               |
+|             **u16** | `65535`                        |           `BigNumberish`            |              `number`               |
+|             **u32** | `4294967295`                   |           `BigNumberish`            |              `number`               |
+|             **u64** | `1.84467E+19 `                 |           `BigNumberish`            |                `BN`                 |
+|             **str** | `anything`                     |              `string`               |              `string`               |
+|            **bool** | `true`                         |              `boolean`              |              `boolean`              |
+|            **b256** | `0x000...`                     |              `string`               |              `string`               |
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/README.md:66]: I will do it, and I also added a note for it here:
 - https://github.com/FuelLabs/fuels-ts/issues/586#issuecomment-1323680625

In doing so, I will need you to merge your PR before I can merge this one.

This is not a problem, just a note to keep in mind.

File: packages/abi-typegen/test/utils/sway/renderTomlTemplate.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    [project]
+    authors = ["fuellabs"]
+    entry = "${params.contractFilename}"
+    license = "Apache-2.0"
+    name = "${params.contractName}"
+
+    [dependencies]`;
+
+  // trim left white spaces
+  return template.replace(/^[\s]+/gm, '');
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/test/utils/sway/renderTomlTemplate.ts:15]: Instead of using the `.replace(/^[\s]+/gm, '')`, you mean?

It already removes all leading white spaces.

File: packages/abi-typegen/test/fixtures/.gitignore [hunk 12]
```diff
@@ -0,0 +1,2 @@
+# Generated files are always ignored
+out
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/test/fixtures/.gitignore:2]: No, the one in the root ignores only the `debug` subfolder with `**/out/debug`, and this one ignores the whole `out` dir _(in this case, there will be things other than a `debug` folder)_.

Are there any rules about where to place these ignore patterns?

I can see multiple files spread all around in various packages.

File: packages/abi-typegen/package.json (hunk 5 above)
Comment (arboleya) [reply packages/abi-typegen/package.json:54]: At first, it wasn't. Now it is.

File: packages/abi-typegen/src/utils/supportedTypes.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+import { TupleType } from '../types/TupleType';
+import { U16Type } from '../types/U16Type';
+import { U32Type } from '../types/U32Type';
+import { U64Type } from '../types/U64Type';
+import { U8Type } from '../types/U8Type';
+import { VectorType } from '../types/VectorType';
+
+export const supportedTypes = [
+  ArrayType,
+  B256Type,
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/src/utils/supportedTypes.ts:20]: Done

File: packages/abi-typegen/test/utils/sway/renderTomlTemplate.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/test/utils/sway/renderTomlTemplate.ts]: Oh, now I see. This was the only place using `replace` for this task — _I made it like the others_.

File: packages/abi-typegen/src/types/U8Type.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (luizstacio) [reply packages/abi-typegen/src/types/U8Type.ts:16]: On the input side, we can accept a BigNumberish that means, string hex, number, or BN. It looks correct. @camsjams we only need to encode different for the return, that looks correct in this case.
Bellow `u32` we return `number`. After `u64` we return a `BN`.

File: packages/abi-typegen/README.md [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+|            **b512** | `fuel1a7r...`                  |              `string`               |              `string`               |
+|          **tuples** | (`MyType`, `MyType`)           |        [`MyType`, `MyType`]         |        [`MyType`, `MyType`]         |
+|           **enums** | enum `MyEnum` { y: (), n: () } | `MyEnum` = `Enum`<{ y: [], n: [] }> | `MyEnum` = `Enum`<{ y: [], n: [] }> |
+|         **structs** | `MyStruct` { a: u8, b: u16 }   |             `MyStruct`              |             `MyStruct`              |
+|         **vectors** | Vec<`MyType`>                  |             `MyType`[]              |             `MyType`[]              |
+|         **options** | Option<`MyType`>               |          `Option<MyType>`           |          `Option<MyType>`           |
+| **raw untyped ptr** | `raw untyped ptr`>             |           `BigNumberish`            |              `number`               |
```
Comment (luizstacio) [inline packages/abi-typegen/README.md:75]:
```suggestion
| **raw untyped ptr** |  `123`             |           `BigNumberish`            |              `number`               |
```

File: packages/abi-typegen/README.md [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+# Type's Conversion Table
+
+The table below describes how Sway types are converted from/to Typescript.
+
+|                Sway | Example                        |              TS:input               |              TS:output              |
+| ------------------: | :----------------------------- | :---------------------------------: | :---------------------------------: |
+|              **u8** | `255`                          |           `BigNumberish`            |              `number`               |
+|             **u16** | `65535`                        |           `BigNumberish`            |              `number`               |
+|             **u32** | `4294967295`                   |           `BigNumberish`            |              `number`               |
+|             **u64** | `1.84467E+19 `                 |           `BigNumberish`            |                `BN`                 |
```
Comment (luizstacio) [inline packages/abi-typegen/README.md:65]:
```suggestion
|             **u64** | `0xFFFFFFFFFFFFFFFF `                 |           `BigNumberish`            |                `BN`                 |
```

File: packages/abi-typegen/README.md [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+```sh
+yarn add @fuel-ts/abi-typegen
+# or
+npm add @fuel-ts/abi-typegen
+```
+
+### Generating types
+
+```sh
+npx typegen -i ./out/debug/*-abi.json -o ./src/contracts
```
Comment (luizstacio) [inline packages/abi-typegen/README.md:35]: Should have fuels on the command?
```suggestion
npx fuels typegen -i ./out/debug/*-abi.json -o ./src/contracts
```

File: packages/abi-typegen/src/utils/parseTypeArguments.ts [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        target,
+        parentTypeId: typeArgument.type,
+        typeArguments: typeArgument.typeArguments,
+      });
+
+      buffer.push(nestedParsed);
+    } else {
+      buffer.push(`${currentLabel}`);
+
+      /*
```
Comment (luizstacio) [inline packages/abi-typegen/src/utils/parseTypeArguments.ts:49]: These comments are some kind of TODO?

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/abi-typegen/src/types/B256Type.ts [hunk 18]
```diff
@@ -0,0 +1,13 @@
+import { StrType } from './StrType';
+
+export class B256Type extends StrType {
+  public static swayTypeExample = 'b256';
+
+  public name = 'b256';
+
+  static MATCH_REGEX = /^b256$/m;
```
Comment (luizstacio) [inline packages/abi-typegen/src/types/B256Type.ts:8]: I didn't understand why having the `swayTypeExample`.

It is just for docs we could add just as docs. I saw it had been used just on `tests`.



```suggestion
  public static swayTypeExample = 'b256';

  public name = 'b256';

  static MATCH_REGEX = /^b256$/m;
```

File: packages/abi-typegen/src/types/B256Type.ts [hunk 19]
```diff
@@ -0,0 +1,13 @@
+import { StrType } from './StrType';
+
+export class B256Type extends StrType {
+  public static swayTypeExample = 'b256';
```
Comment (luizstacio) [inline packages/abi-typegen/src/types/B256Type.ts:4]: I didn't understand why having the `swayTypeExample`.

It is just for docs we could add just as docs. I saw it had been used just on `tests`.

File: packages/abi-typegen/README.md (hunk 16 above)
Comment (arboleya) [reply packages/abi-typegen/README.md:35]: No, only when choosing the full SDK installation.

File: packages/abi-typegen/src/utils/parseTypeArguments.ts (hunk 17 above)
Comment (arboleya) [reply packages/abi-typegen/src/utils/parseTypeArguments.ts:49]: No, just some annotations I don't want to risk losing just yet.

Planning on doing so right before (or _after_) the merge.

File: packages/abi-typegen/README.md (hunk 16 above)
Comment (arboleya) [reply packages/abi-typegen/README.md:35]: At this point, the `fuels` prefix is required only when using the umbrella package.

I will, however, rename it from `typegen` to `fuels-typegen` (_only for the standalone version).

> _Same approach I used for the `fuels version` x `fuels-version` command from #616_.

File: packages/abi-typegen/src/templates/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/src/templates/index.ts:13]: I will integrate this on top of #616, and then annotate the generated files with versions for everything.

File: packages/abi-typegen/src/types/B256Type.ts (hunk 19 above)
Comment (luizstacio) [reply packages/abi-typegen/src/types/B256Type.ts:4]: Yeah maybe removing the example would help

File: packages/abi-typegen/src/utils/parseTypeArguments.ts (hunk 17 above)
Comment (luizstacio) [reply packages/abi-typegen/src/utils/parseTypeArguments.ts:49]: maybe it should be removed now as the PR is ready to review/merge...

File: packages/abi-typegen/src/utils/getPackageVersion.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/src/utils/getPackageVersion.test.ts:8]: > **Note**: I changed the root `package.json:version` to `0.0.0`

File: packages/abi-typegen/test/fixtures/array-nested.sw (hunk 6 above)
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-typegen/test/fixtures/array-nested.sw]: Guess this is it: https://github.com/FuelLabs/fuels-ts/pull/570/commits/bab74003a281dc459b1b962bce15ad2a46d8d625

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/fuels/src/index.ts [hunk 20]
```diff
@@ -1,5 +1,6 @@
 export * from '@ethersproject/bytes';
 export * from '@fuel-ts/abi-coder';
+export * from '@fuel-ts/abi-typegen';
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/fuels/src/index.ts:3]: You are totally right, thanks.

It is _safe_ but not useful enough to be exported by default.

File: packages/abi-typegen/src/index.test.ts [hunk 21]
```diff
@@ -0,0 +1,7 @@
+import * as indexMod from './AbiTypeGen';
+
+describe('index.ts', () => {
+  test('should export AbiTypeGen class', () => {
+    expect(indexMod.AbiTypeGen).toBeTruthy;
```
Comment (luizstacio) [inline packages/abi-typegen/src/index.test.ts:5]: it should use be `toBeTruthy()`?

File: packages/abi-typegen/src/templates/dts.ts [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+      });
+    });
+
+  // Handles custom common types
+  let commonImports = renderDtsCommonImportTemplate({ commonTypesInUse });
+
+  commonImports = commonTypesInUse.length ? `\n${commonImports}\n` : '';
+
+  // glue everything
+  const output = `/* Autogenerated file. Do not edit manually. */
```
Comment (luizstacio) [inline packages/abi-typegen/src/templates/dts.ts:127]: We could replace these templates by placing them on files and using `handlebars` or another template processor to replace the dynamic parts.

File: packages/abi-typegen/src/templates/dts.test.ts [hunk 23]
```diff
@@ -0,0 +1,96 @@
+import { contractPaths } from '../../test/fixtures';
+import { compileSwayToJson } from '../../test/utils/sway/compileSwayToJson';
+import { Abi } from '../Abi';
+import { getPackageVersion } from '../utils/getPackageVersion';
+
+import { renderDtsTemplate } from './dts';
+
+const expectedRenderedTemplate = `/* Autogenerated file. Do not edit manually. */
```
Comment (luizstacio) [inline packages/abi-typegen/src/templates/dts.test.ts:8]: this template specs could be placed on files under folder specs or mocks maybe.

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/abi-typegen/src/cli.ts [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+  log('Generating files..\n');
+
+  mkdirp.sync(`${outputDir}/factories`);
+
+  abiTypeGen.files.forEach((file) => {
+    rimraf.sync(file.path);
+    writeFileSync(file.path, file.contents);
+    const trimPathRegex = new RegExp(`^.+${cwdBasename}/`, 'm');
+    log(` - ${file.path.replace(trimPathRegex, '')}`);
+  });
```
Comment (luizstacio) [inline packages/abi-typegen/src/cli.ts:85]: We should extract this part to another file enabling it to run typegen programmatically.

Example of implementation using `fuelchain`
```ts
import { runTypeChain } from 'fuelchain';

await runTypeChain({
    cwd,
    filesToProcess: ['<contract path 1>', '<contract path 1>'],
    allFiles,
    outDir: config.types.output,
    target: 'fuels',
});
```

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/abi-typegen/src/index.test.ts (hunk 21 above)
Comment (arboleya) [reply packages/abi-typegen/src/index.test.ts]: Nice catch!

File: packages/abi-typegen/src/types/GenericType.ts [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+
+  static MATCH_REGEX: RegExp = /^generic ([^\s]+)$/m;
+
+  static isSuitableFor(params: { type: string }) {
+    return GenericType.MATCH_REGEX.test(params.type);
+  }
+
+  public parseComponentsAttributes(_params: { types: IType[] }) {
+    const matched = this.rawAbiType.type.match(GenericType.MATCH_REGEX);
+    const label = matched![1];
```
Comment (luizstacio) [inline packages/abi-typegen/src/types/GenericType.ts:18]: It would be good to check here and throw a descriptive error if the type doesn't match. Or if is really not needed add eslint ignore comment

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: .vscode/settings.json [hunk 26]
```diff
@@ -18,5 +18,16 @@
...
     "editor.defaultFormatter": "esbenp.prettier-vscode"
+  },
+  "[handlebars]": {
+    "editor.formatOnSave": false,
+    "editor.formatOnPaste": false
+  },
+  "path-autocomplete.includeExtension": true,
+  "path-autocomplete.extensionOnImport": true,
+  "path-autocomplete.excludedItems": {
+    // do not mess with defaut auto-complete for `js` and `ts`
```
Comment (luizstacio) [inline .vscode/settings.json:29]: I think comments inside the JSON are not a valid syntax.

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: .vscode/settings.json (hunk 26 above)
Comment (arboleya) [reply .vscode/settings.json:29]: https://code.visualstudio.com/Docs/languages/json#_json-with-comments

File: packages/abi-typegen/README.md [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+```sh
+npx fuels typegen -i ./out/debug/*-abi.json -o ./src/contracts
+```
+
+## Programatic API
+
+```ts
+import { runTypegen } from "@fuel-ts/abi-typegen";
+// OR
+// import { runTypegen } from 'fuels'
```
Comment (luizstacio) [inline packages/abi-typegen/README.md:65]: I think this is not possible once the `@fuel-ts/abi-typegen` is not exported on `fuels` anymore.

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/abi-typegen/README.md (hunk 27 above)
Comment (arboleya) [reply packages/abi-typegen/README.md:65]: So true! 🦅

**Review by luizstacio — APPROVED** [review]

**Review by luizstacio — DISMISSED** [review]

**Review by luizstacio — DISMISSED** [review]

**Review by Dhaiwat10 — DISMISSED** [review]
That's one hell of a PR 😅 Enjoyed going through all the code and reading it

**Review by Dhaiwat10 — APPROVED** [review]

**Review by luizstacio — APPROVED** [review]
